src/ucm/bistro/bistro_x86_64.c

```diff
@@ -14,19 +14,153 @@
 #include <stdlib.h>
+#include <pthread.h>
 
 #include <ucm/bistro/bistro.h>
 #include <ucm/bistro/bistro_int.h>
 #include <ucm/util/sys.h>
 #include <ucs/sys/math.h>
-#include <ucs/arch/cpu.h>
-#include <ucs/debug/assert.h>
 
 
-ucs_status_t ucm_bistro_patch(const char *symbol, void *hook,
+typedef struct {
+    void    *jmp_address;
+    char    code[];
+} ucm_bistro_orig_func_t;
+
+typedef struct {
+    uint8_t  opcode; /* 0xff */
+    uint8_t  modrm;  /* 0x25 */
+    int32_t  displ;
+} UCS_S_PACKED ucm_bistro_jmp_indirect_t;
+
+
+/*
+ * Find the minimal length of initial instructions in the function which can be
+ * safely executed from any memory location.
+ * Uses a very simplified disassembler which supports only the typical
+ * instructions found in function prologue.
+ */
+static size_t ucm_bistro_find_movable_prefix(const void *func, size_t min_length)
+{
+    size_t offset, prev_offset;
+    uint8_t rex, opcode, modrm;
+
+    offset = 0;
+    while (offset < min_length) {
+        prev_offset = offset;
+        opcode      = *(uint8_t*)UCS_PTR_BYTE_OFFSET(func, offset++);
+
+        /* check for REX prefix */
+        if ((opcode & 0xF0) == 0x40) {
+            rex    = opcode;
+            opcode = *(uint8_t*)UCS_PTR_BYTE_OFFSET(func, offset++);
+        } else {
+            rex = 0;
+        }
+
+        /* check the opcode */
+        if (((rex == 0) || rex == 0x41) && ((opcode & 0xF0) == 0x50)) {
+            /* push <register> */
+            continue;
+        } else if ((rex == 0x48) && (opcode == 0x81)) {
+            /* group 1A operation */
+            modrm = *(uint8_t*)UCS_PTR_BYTE_OFFSET(func, offset++);
+            if (modrm == 0xEC) {
+                /* sub $imm32, %rsp */
+                offset += sizeof(uint32_t);
+                continue;
+            }
+        } else if ((rex == 0x48) && (opcode == 0x89)) {
+            /* mov %rsp, %rbp */
+            modrm = *(uint8_t*)UCS_PTR_BYTE_OFFSET(func, offset++);
+            if (modrm == 0xE5) {
+                continue;
+            }
+        }
+
+        /* unsupported instruction - bail */
+        return prev_offset;
+    }
+
+    return offset;
+}
+
+static void *ucm_bistro_allocate_code(size_t size)
+{
+    static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+    void *page                   = MAP_FAILED;
+    size_t offset                = 0;
+    void *result;
+
+    pthread_mutex_lock(&mutex);
+
+    if (page == MAP_FAILED) {
+        /* allocate executable page */
+        page = mmap(NULL, ucm_get_page_size(),
+                    PROT_READ | PROT_WRITE | PROT_EXEC,
+                    MAP_PRIVATE | MAP_ANONYMOUS, 0, 0);
+        if (page == MAP_FAILED) {
+            result = NULL;
+            goto out;
+        }
+    }
+
+    if (size + offset > ucm_get_page_size()) {
+        result = NULL;
+        goto out;
+    }
+
+    /* allocate space in the page */
+    result  = UCS_PTR_BYTE_OFFSET(page, offset);
+    offset += size;
```

Comment:
it seems body of this function was in loop :)

Reply:
forgot to add 'static'